File: java/src/main/java/com/pahappa/systems/commssdk/v1/CommsSDK.java
package com.pahappa.systems.commssdk.v1;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.pahappa.systems.commssdk.v1.models.*;
import com.pahappa.systems.commssdk.v1.utils.NumberValidator;
import com.pahappa.systems.commssdk.v1.utils.Validator;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import lombok.Getter;
import lombok.Setter;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

/**
 * Main entry point for interacting with the CommsSDK.
 * <p>
 * Provides methods for authentication, sending SMS, querying balances, and environment configuration.
 * </p>
 */
public class CommsSDK {

    /**
     * Shared Jackson object mapper for JSON serialization.
     */
    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    /**
     * The API endpoint URL. Defaults to the live server.
     */
    public static String API_URL = "https://comms.egosms.co/api/v1/json/";

    @Getter
    private String userName;

    @Getter
    private String apiKey;

    @Getter
    @Setter
    private String senderId = "EgoSMS";

    @Setter
    private boolean isAuthenticated = false;

    private final RestTemplate client = new RestTemplate();

    /**
     * Private constructor. Use {@link #authenticate(String, String)} to create an instance.
     */
    private CommsSDK() {
    }

    /**
     * Authenticates and creates a new CommsSDK instance.
     *
     * @param userName Your account username.
     * @param apiKey   Your API key.
     * @return Authenticated CommsSDK instance.
     */
    public static CommsSDK authenticate(String userName, String apiKey) {
        CommsSDK sdk = new CommsSDK();
        sdk.userName = userName;
        sdk.apiKey = apiKey;
        Validator.validateCredentials(sdk);
        return sdk;
    }

    /**
     * Switches the SDK to use the sandbox environment (for testing).
     * <br>
     * Make an account at <a href="http://comms-test.pahappa.net">comms-test.pahappa.net</a> to use the sandbox.
     * Use {@link CommsSDK#useLiveServer()} for the live server.
     */
    public static void useSandBox() {
        API_URL = "https://comms-test.pahappa.net/api/v1/json";
    }

    /**
     * Switches the SDK to use the live environment (for production).
     * <br>
     * Make an account at <a href="http://comms.egosms.co">comms.egosms.co</a> to use the live server.
     * Use {@link CommsSDK#useSandBox()} for the sandbox server.
     */
    public static void useLiveServer() {
        API_URL = "https://comms.egosms.co/api/v1/json";
    }

    /**
     * Marks the SDK as authenticated.
     * <b>Not to be called manually!</b>
     */
    public void setAuthenticated() {
        this.isAuthenticated = true;
    }

    /**
     * Sets the sender ID for outgoing messages.
     *
     * @param senderId Sender ID (max 11 characters).
     * @return This CommsSDK instance (for chaining).
     */
    public CommsSDK withSenderId(String senderId) {
        this.senderId = senderId;
        return this;
    }

    /**
     * Sends an SMS to a single number with default sender ID and highest priority.
     *
     * @param number  Recipient phone number.
     * @param message Message text.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(String number, String message) {
        return sendSMS(
                Collections.singletonList(number),
                message,
                senderId,
                MessagePriority.HIGHEST
        );
    }

    /**
     * Sends an SMS to a single number with custom sender ID.
     *
     * @param number   Recipient phone number.
     * @param message  Message text.
     * @param senderId Sender ID.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(String number, String message, String senderId) {
        return sendSMS(
                Collections.singletonList(number),
                message,
                senderId,
                MessagePriority.HIGHEST
        );
    }

    /**
     * Sends an SMS to a single number with custom sender ID and priority.
     *
     * @param number   Recipient phone number.
     * @param message  Message text.
     * @param senderId Sender ID.
     * @param priority Message priority.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(
            String number,
            String message,
            String senderId,
            MessagePriority priority
    ) {
        return sendSMS(
                Collections.singletonList(number),
                message,
                senderId,
                priority
        );
    }

    /**
     * Sends an SMS to a single number with default sender ID and custom priority.
     *
     * @param number   Recipient phone number.
     * @param message  Message text.
     * @param priority Message priority.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(
            String number,
            String message,
            MessagePriority priority
    ) {
        return sendSMS(
                Collections.singletonList(number),
                message,
                senderId,
                priority
        );
    }

    /**
     * Sends an SMS to multiple numbers with default sender ID and highest priority.
     *
     * @param numbers List of recipient phone numbers.
     * @param message Message text.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(List<String> numbers, String message) {
        return sendSMS(numbers, message, senderId, MessagePriority.HIGHEST);
    }

    /**
     * Sends an SMS to multiple numbers with custom sender ID.
     *
     * @param numbers  List of recipient phone numbers.
     * @param message  Message text.
     * @param senderId Sender ID.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(
            List<String> numbers,
            String message,
            String senderId
    ) {
        return sendSMS(numbers, message, senderId, MessagePriority.HIGHEST);
    }

    /**
     * Sends an SMS to multiple numbers with default sender ID and custom priority.
     *
     * @param numbers  List of recipient phone numbers.
     * @param message  Message text.
     * @param priority Message priority.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(
            List<String> numbers,
            String message,
            MessagePriority priority
    ) {
        return sendSMS(numbers, message, senderId, priority);
    }

    /**
     * Sends an SMS to multiple numbers with custom sender ID and priority.
     *
     * @param numbers  List of recipient phone numbers.
     * @param message  Message text.
     * @param senderId Sender ID.
     * @param priority Message priority.
     * @return true if sent successfully, false otherwise.
     */
    public boolean sendSMS(
            List<String> numbers,
            String message,
            String senderId,
            MessagePriority priority
    ) {
        ApiResponse apiResponse = querySendSMS(
                numbers,
                message,
                senderId,
                priority
        );
        if (apiResponse == null) {
            System.out.println("Failed to get a response from the server.");
            return false;
        }
        switch (apiResponse.getStatus()) {
            case OK:
                System.out.println("SMS sent successfully.");
                System.out.println(
                        "MessageFollowUpUniqueCode: " + apiResponse.getMessageFollowUpCode()
                );
                return true;
            case Failed:
                System.out.println("Failed: " + apiResponse.getMessage());
                return false;
            default:
                throw new RuntimeException(
                        "Unexpected response status: " + apiResponse.getStatus()
                );
        }
    }

    /**
     * Sends an SMS and returns the full API response object.
     *
     * @param numbers  List of recipient phone numbers.
     * @param message  Message text.
     * @param senderId Sender ID.
     * @param priority Message priority.
     * @return ApiResponse object with status and details, or null on error.
     */
    public ApiResponse querySendSMS(
            List<String> numbers,
            String message,
            String senderId,
            MessagePriority priority
    ) {
        if (sdkNotAuthenticated()) return null;
        if (numbers == null || numbers.isEmpty()) {
            throw new IllegalArgumentException("Numbers list cannot be empty");
        }
        if (message == null || message.isEmpty()) {
            throw new IllegalArgumentException("Message cannot be empty");
        }
        if (message.length() == 1) {
            throw new IllegalArgumentException(
                    "Message cannot be a single character"
            );
        }
        if (senderId == null || senderId.trim().isEmpty()) {
            senderId = this.senderId;
        }
        if (senderId != null && senderId.length() > 11) {
            System.out.println(
                    "Warning: Sender ID length exceeds 11 characters. Some networks may truncate or reject messages."
            );
        }
        if (priority == null) {
            priority = MessagePriority.HIGHEST;
        }
        numbers = NumberValidator.validateNumbers(numbers);
        if (numbers.isEmpty()) {
            System.err.println(
                    "No valid phone numbers provided. Please check inputs."
            );
            return null;
        }
        ApiRequest apiRequest = new ApiRequest();
        apiRequest.setMethod("SendSms");
        List<MessageModel> messageModels = new ArrayList<>();
        for (String num : numbers) {
            MessageModel messageModel = new MessageModel();
            messageModel.setNumber(num);
            messageModel.setMessage(message);
            messageModel.setSenderId(senderId);
            messageModel.setPriority(priority);
            messageModels.add(messageModel);
        }
        apiRequest.setMessageData(messageModels);
        apiRequest.setUserdata(new UserData(userName, apiKey));
        ResponseEntity<String> res = client.postForEntity(
                API_URL,
                apiRequest,
                String.class
        );
        try {
            return OBJECT_MAPPER.readValue(res.getBody(), ApiResponse.class);
        } catch (Exception e) {
            System.err.println("Failed to send SMS: " + e.getMessage());
            try {
                System.err.println(
                        "Request: " + OBJECT_MAPPER.writeValueAsString(apiRequest)
                );
            } catch (Exception ignored) {
            }
            return null;
        }
    }

    /**
     * Checks if the SDK is authenticated. If not, attempts to re-authenticate.
     *
     * @return true if not authenticated, false if authenticated.
     */
    private boolean sdkNotAuthenticated() {
        if (!isAuthenticated) {
            System.err.println(
                    "SDK is not authenticated. Please authenticate before performing actions."
            );
            System.err.println(
                    "Attempting to re-authenticate with provided credentials..."
            );
            return !Validator.validateCredentials(this);
        }
        return false;
    }

    /**
     * Queries the balance and returns the full API response object.
     *
     * @return ApiResponse object with balance and details, or null on error.
     */
    public ApiResponse queryBalance() {
        if (sdkNotAuthenticated()) {
            return null;
        }
        ApiRequest apiRequest = new ApiRequest();
        apiRequest.setMethod("Balance");
        apiRequest.setUserdata(new UserData(userName, apiKey));
        try {
            ResponseEntity<String> res = client.postForEntity(
                    API_URL,
                    apiRequest,
                    String.class
            );
            return OBJECT_MAPPER.readValue(res.getBody(), ApiResponse.class);
        } catch (Exception e) {
            throw new RuntimeException("Failed to get balance: " + e.getMessage(), e);
        }
    }

    /**
     * Gets your current SMS account balance.
     *
     * @return Balance as a double.
     */
    public double getBalance() {
        return queryBalance().getBalance();
    }

    /**
     * Returns a string representation of the SDK instance.
     *
     * @return String representation.
     */
    @Override
    public String toString() {
        return "SDK(" + userName + " => " + apiKey + ")";
    }
}

File: ruby/lib/comms_sdk/v1/comms_sdk.rb
# frozen_string_literal: true

require 'net/http'
require 'json'
require 'uri'
require_relative 'models'
require_relative 'utils'

module CommsSdk
  module V1
    class CommsSDK
      API_URL = "http://176.58.101.43:8080/communications/api/v1/json/"

      attr_reader :api_key, :user_name, :sender_id, :is_authenticated

      def initialize
        @api_key = nil
        @user_name = nil
        @sender_id = "EgoSMS"
        @is_authenticated = false
      end

      def self.authenticate(user_name, api_key)
        sdk = new
        sdk.instance_variable_set(:@user_name, user_name)
        sdk.instance_variable_set(:@api_key, api_key)
        Utils::Validator.validate_credentials(sdk)
        sdk
      end

      def self.use_sandbox
        remove_const(:API_URL) if const_defined?(:API_URL)
        const_set(:API_URL, "http://176.58.101.43:8080/communications/api/v1/json")
      end

      def self.use_live_server
        remove_const(:API_URL) if const_defined?(:API_URL)
        const_set(:API_URL, "http://176.58.101.43:8080/communications/api/v1/json")
      end

      def set_authenticated
        @is_authenticated = true
      end

      def with_sender_id(sender_id)
        @sender_id = sender_id
        self
      end

      def send_sms(numbers, message, sender_id: nil, priority: MessagePriority::HIGHEST)
        numbers = [numbers] if numbers.is_a?(String)
        
        api_response = query_send_sms(numbers, message, sender_id || @sender_id, priority)
        
        if api_response.nil?
          puts "Failed to get a response from the server."
          return false
        end
        
        case api_response.status
        when ApiResponseCode::OK
          puts "SMS sent successfully."
          puts "MessageFollowUpUniqueCode: #{api_response.msg_follow_up_unique_code}" if api_response.msg_follow_up_unique_code
          true
        when ApiResponseCode::FAILED
          puts "Failed: #{api_response.message}"
          false
        else
          raise "Unexpected response status: #{api_response.status}"
        end
      end

      def query_send_sms(numbers, message, sender_id, priority)
        return nil if sdk_not_authenticated?
        
        raise ArgumentError, "Numbers list cannot be empty" if numbers.nil? || numbers.empty?
        raise ArgumentError, "Message cannot be empty" if message.nil? || message.empty?
        raise ArgumentError, "Message cannot be a single character" if message.length == 1
        
        sender_id = @sender_id if sender_id.nil? || sender_id.strip.empty?
        
        if sender_id.length > 11
          puts "Warning: Sender ID length exceeds 11 characters. Some networks may truncate or reject messages."
        end
        
        numbers = Utils::NumberValidator.validate_numbers(numbers)
        if numbers.empty?
          $stderr.puts "No valid phone numbers provided. Please check inputs."
          return nil
        end
        
        message_models = numbers.map do |num|
          MessageModel.new(number: num, message: message, senderid: sender_id, priority: priority.value)
        end
        
        api_request = ApiRequest.new(
          method: "SendSms",
          userdata: UserData.new(@user_name, @api_key),
          msgdata: message_models
        )
        
        begin
          response = make_http_request(api_request)
          ApiResponse.from_hash(JSON.parse(response.body))
        rescue => e
          $stderr.puts "Failed to send SMS: #{e.message}"
          begin
            $stderr.puts "Request: #{api_request.to_hash}"
          rescue
            # Ignore serialization errors
          end
          nil
        end
      end

      def query_balance
        return nil if sdk_not_authenticated?
        
        api_request = ApiRequest.new(
          method: "Balance",
          userdata: UserData.new(@user_name, @api_key)
        )
        
        begin
          response = make_http_request(api_request)
          ApiResponse.from_hash(JSON.parse(response.body))
        rescue => e
          raise "Failed to get balance: #{e.message}"
        end
      end

      def get_balance
        response = query_balance
        return nil if response.nil? || response.balance.nil?
        
        begin
          Float(response.balance)
        rescue ArgumentError
          nil
        end
      end

      def to_s
        "SDK(#{@user_name} => #{@api_key})"
      end

      private

      def sdk_not_authenticated?
        unless @is_authenticated
          $stderr.puts "SDK is not authenticated. Please authenticate before performing actions."
          $stderr.puts "Attempting to re-authenticate with provided credentials..."
          return !Utils::Validator.validate_credentials(self)
        end
        false
      end

      def make_http_request(api_request)
        uri = URI(API_URL)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = uri.scheme == 'https'
        
        request = Net::HTTP::Post.new(uri)
        request['Content-Type'] = 'application/json'
        request.body = api_request.to_json
        
        http.request(request)
      end
    end
  end
end